### DISCUSSION TITLE
Make BlockSapling tick logic more efficient

### DISCUSSION THREAD

**Discussion by bruberu** [ACGaming/UniversalTweaks#300] [Tweak Suggestions]
While looking through BlockSapling's code, I noticed this function:
``` 
    public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
    {
        if (!worldIn.isRemote)
        {
            super.updateTick(worldIn, pos, state, rand);

            if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
            if (worldIn.getLightFromNeighbors(pos.up()) >= 9 && rand.nextInt(7) == 0)
            {
                this.grow(worldIn, pos, state, rand);
            }
        }
    }
```
Looking further, I noticed that isAreaLoaded is an O(n^2) algorithm, which presumably would take up a decent amount of tick time if enough saplings were placed down (e.g. in a skyblock pack). As such, UniversalTweaks could move the random check just before the isAreaLoaded call (not before the super call though, since that makes sure the sapling is still on the ground). In fact, this mod could also create a configuration to replace the magic number 7 here, allowing pack creators to speed up sapling growth if they so desire.

**Comment by ACGaming** [comment]
Closed with 09c43e4e8a4376a69616c91a42a2c4cd5dc5e6f4

**Reply by jchung01** [reply]
As per the Forge comment, you probably still want to call `getLightFromNeighbors()` after `isAreaLoaded` to avoid unnecessary chunk loading.

**Reply by ACGaming** [reply]
Right, this should be the best order.
